airbyte-integrations/bases/base-python/base_python/sdk/abstract_source.py

```diff
@@ -0,0 +1,148 @@
+
+    def discover(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteCatalog:
+        """Discover streams"""
+        streams = []
+
+        for name, stream in self.streams(config=config).items():
+            args = {
+                'name': name,
+                'json_schema': stream.get_json_schema(),
+            }
+
+            supported_sync_modes = [SyncMode.full_refresh]
+            if isinstance(stream, IncrementalStream):
+                supported_sync_modes.append(SyncMode.incremental)
+                args['source_defined_cursor'] = stream.source_defined_cursor
+                args['default_cursor_field'] = [stream.cursor_field] if isinstance(stream.cursor_field, str) else stream.cursor_field
+
+            args['supported_sync_modes'] = supported_sync_modes
+
+            streams.append(AirbyteStream(**args))
+
+        return AirbyteCatalog(streams=streams)
+
+    def check(self, logger: AirbyteLogger, config: Mapping[str, Any]) -> AirbyteConnectionStatus:
+        """Check connection"""
+        alive, error = self.check_connection(logger, config)
+        if not alive:
+            return AirbyteConnectionStatus(status=Status.FAILED, message=str(error))
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def read(
+            self, logger: AirbyteLogger, config: Mapping[str, Any], catalog: ConfiguredAirbyteCatalog, state: MutableMapping[str, Any] = None
+    ) -> Iterator[AirbyteMessage]:
+
+        state = state or {}
+        total_state = copy.deepcopy(state)
+        logger.info(f"Starting syncing {self.name}")
+        # TODO assert all streams exist in the connector
+        for configured_stream in catalog.streams:
+            try:
+                yield from self._read_stream(logger=logger, config=config, configured_stream=configured_stream, state=total_state)
+            except Exception as e:
+                logger.exception(f"Encountered an exception while reading stream {self.name}")
+                raise e
+
+        logger.info(f"Finished syncing {self.name}")
+
+    def _read_stream(
+            self, logger: AirbyteLogger, config: Mapping[str, Any], configured_stream: ConfiguredAirbyteStream, state: MutableMapping[str, Any]
+    ) -> Iterator[AirbyteMessage]:
+        stream_name = configured_stream.stream.name
+        stream_instance = self.streams(config)[stream_name]
+        use_incremental = configured_stream.sync_mode == SyncMode.incremental and isinstance(stream_instance, IncrementalStream)
+
+        stream_state = {}
+        if use_incremental and state.get(stream_name):
+            logger.info(f"Set state of {stream_name} stream to {state.get(stream_name)}")
+            stream_state = state.get(stream_name)
+
+        logger.info(f"Syncing {stream_name} stream")
+        record_counter = 0
+        for record in stream_instance.read_stream(stream_state=copy.deepcopy(stream_state)):
+            now = int(datetime.now().timestamp()) * 1000
```

Comment:
can you call it now_millis or something like that?

Reply:
done